server/android_notification_server.go

```diff
@@ -272,26 +272,31 @@ func (me *AndroidNotificationServer) SendNotificationWithRetry(fcmMsg *messaging
 
 		select {
 		case <-generalContext.Done():
+			if generalContext.Err() != nil {
+				logger.Info(
+					"Not retrying because context error",
+					mlog.Int("retry", retries),
+					mlog.Err(generalContext.Err()),
+				)
+			}
+			return generalContext.Err()
 		case <-time.After(waitTime):
 		}
 
-		if generalContext.Err() != nil {
-			logger.Info(
-				"Not retrying because context error",
-				mlog.Int("retry", retries),
-				mlog.Err(generalContext.Err()),
-			)
-			err = generalContext.Err()
-			break
-		}
-
 		waitTime *= 2
 	}
 
 	return err
 }
 
 func isRetryable(err error) bool {
+	// We retry if the context deadline is exceeded.
+	// This may cause double notifications, but we expect
+	// this not to happen often.
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
```

Comment:
If it does start happening more often than expected, do we have a way of tracking this?

Reply:
Apart of the bug reports from notifications arriving more than once, on line 263 we are logging all errors that imply a retry.